pkg/controller/egressgateway/egressgateway_controller.go

```diff
@@ -0,0 +1,666 @@
+		Installation:      installation,
+		OSType:            rmeta.OSTypeLinux,
+		EgressGW:          egw,
+		EgressGWVxlanPort: egwVxlanPort,
+		EgressGWVxlanVNI:  egwVxlanVNI,
+		Openshift:         openshift,
+		UsePSP:            r.usePSP,
+	}
+
+	component := egressgateway.EgressGateway(config)
+	ch := utils.NewComponentHandler(log, r.client, r.scheme, egw)
+
+	if err = imageset.ApplyImageSet(ctx, r.client, variant, component); err != nil {
+		reqLogger.Error(err, "Error with images from ImageSet")
+		r.status.SetDegraded(operatorv1.ResourceUpdateError, "Error with images from ImageSet", err, reqLogger)
+		setDegraded(r.client, ctx, egw, reconcileErr,
+			fmt.Sprintf("Error with images from ImageSet Name = %s, Namespace = %s, err = %s", egw.Name, egw.Namespace, err.Error()))
+		return reconcile.Result{}, err
+	}
+
+	if err = ch.CreateOrUpdateOrDelete(ctx, component, r.status); err != nil {
+		reqLogger.Error(err, fmt.Sprintf("Error creating / updating resource: Name = %s, Namespace = %s", egw.Name, egw.Namespace))
+		r.status.SetDegraded(operatorv1.ResourceUpdateError,
+			fmt.Sprintf("Error creating / updating resource: Name = %s, Namespace = %s", egw.Name, egw.Namespace), err, reqLogger)
+		setDegraded(r.client, ctx, egw, reconcileErr,
+			fmt.Sprintf("Error creating / updating resource Name = %s, Namespace = %s, err = %s", egw.Name, egw.Namespace, err.Error()))
+		return reconcile.Result{}, err
+	}
+
+	// Update the status of this CR.
+	egw.Status.State = operatorv1.TigeraStatusReady
+	setAvailable(r.client, ctx, egw, string(operatorv1.AllObjectsAvailable), "All objects available")
+
+	// After the resource is created/updated, TigeraStatus needs to be set by taking the cumulative status of the
+	// available EGW resources. Lets say we create 2 EGW resources Red, Blue. Both are degraded. Now lets create 3rd resource
+	// yellow. Though yellow is reconciled and rendered successfully, TigeraStatus should still be degraded as Red, Blue are
+	// degraded. Now lets assume Blue gets updated and gets rendered properly. In this case, TigeraStatus should still be
+	// degraded as Red is unhealthy. "status" represents, the cumulative status of EGWs.
+	if unreadyEGW != nil {
+		r.status.SetDegraded(operatorv1.ResourceCreateError,
+			fmt.Sprintf("Error reconciling Egress Gateway resource. Name=%s Namespace=%s", unreadyEGW.Name, unreadyEGW.Namespace),
+			fmt.Errorf("%s", getDegradedMsg(unreadyEGW)), reqLogger)
+		return reconcile.Result{}, nil
+	}
+
+	r.status.ClearDegraded()
+	if !r.status.IsAvailable() {
+		// Schedule a kick to check again in the near future, hopefully by then things will be available.
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+	return reconcile.Result{}, nil
+}
+
+// getRequestedEgressGateway returns the namespaced EgressGateway instance.
+func getRequestedEgressGateway(egws []operatorv1.EgressGateway, request reconcile.Request) (*operatorv1.EgressGateway, int) {
+	for index, egw := range egws {
```

Comment:
Why do we iterate over the list rather than just querying it using the client? 

The client uses a cache, so it won't generate a new request unless it needs to. We should probably refactor this to use the client unless there's a reason I'm not seeing.

Reply:
Since all the status management is done here, I need the index of the requested resource. Hence I did not use the client.